hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/realtime/HoodieMergeOnReadSnapshotReader.java

```diff
@@ -0,0 +1,216 @@
+import static org.apache.hudi.hadoop.config.HoodieRealtimeConfig.DEFAULT_MAX_DFS_STREAM_BUFFER_SIZE;
+import static org.apache.hudi.hadoop.config.HoodieRealtimeConfig.DEFAULT_SPILLABLE_MAP_BASE_PATH;
+import static org.apache.hudi.hadoop.config.HoodieRealtimeConfig.ENABLE_OPTIMIZED_LOG_BLOCKS_SCAN;
+import static org.apache.hudi.hadoop.config.HoodieRealtimeConfig.MAX_DFS_STREAM_BUFFER_SIZE_PROP;
+import static org.apache.hudi.hadoop.config.HoodieRealtimeConfig.SPILLABLE_MAP_BASE_PATH_PROP;
+import static org.apache.hudi.hadoop.utils.HoodieRealtimeRecordReaderUtils.getBaseFileReader;
+import static org.apache.hudi.hadoop.utils.HoodieRealtimeRecordReaderUtils.getMaxCompactionMemoryInBytes;
+import static org.apache.hudi.internal.schema.InternalSchema.getEmptyInternalSchema;
+
+/**
+ * An implementation of {@link AbstractRealtimeRecordReader} that reads from base parquet files and log files,
+ * and merges the records on the fly. It differs from {@link HoodieRealtimeRecordReader} in that it does not
+ * implement Hadoop's RecordReader interface, and instead implements Iterator interface that returns an iterator
+ * of {@link HoodieRecord}s which are {@link HoodieAvroIndexedRecord}s. This can be used by query engines like
+ * Trino that do not use Hadoop's RecordReader interface. However, the engine must support reading from iterators
+ * and also support Avro (de)serialization.
+ */
+public class HoodieMergeOnReadSnapshotReader extends AbstractRealtimeRecordReader implements Iterator<HoodieRecord>, AutoCloseable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HoodieMergeOnReadSnapshotReader.class);
+
+  private final String tableBasePath;
+  private final List<HoodieLogFile> logFilePaths;
+  private final String latestInstantTime;
+  private final Schema readerSchema;
+  private final JobConf jobConf;
+  private final HoodieMergedLogRecordScanner logRecordScanner;
+  private final HoodieFileReader baseFileReader;
+  private final Map<String, HoodieRecord> logRecordsByKey;
+  private final Iterator<HoodieRecord> recordsIterator;
+  private final ExternalSpillableMap<String, HoodieRecord> mergedRecordsByKey;
+
+  /**
+   * In order to instantiate this record reader, one needs to provide following parameters.
+   * An example usage is demonstrated in TestHoodieMergeOnReadSnapshotReader.
+   *
+   * @param tableBasePath     Base path of the Hudi table
+   * @param baseFilePath      Path of the base file as of the latest instant time for the split being processed
+   * @param logFilePaths      Paths of the log files as of the latest file slices pertaining to file group id of the base file
+   * @param latestInstantTime Latest instant time
+   * @param readerSchema      Schema of the reader
+   * @param jobConf           Any job configuration
+   * @param start             Start offset
+   * @param length            Length of the split
+   */
+  public HoodieMergeOnReadSnapshotReader(String tableBasePath,
```

Comment:
oh man. another record reader impl? 
Did we take a look at [HoodieMergedReadHandle](https://github.com/apache/hudi/blob/0ca0999420ca36b777b23fc9cf3998ce8f31481f/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieMergedReadHandle.java) ?